Nef_3/include/CGAL/Nef_3/K3_tree.h

```diff
@@ -664,163 +583,153 @@ std::string dump_object_list( const Object_list& O, int level = 0) {
+  Side_of_plane sop(point_on_plane, coord, reference_counted);
 #else
-  Side_of_plane sop;
+  Side_of_plane sop(point_on_plane, coord);
 #endif
-  typename Object_list::size_type v_end1 = O1.size();
-  typename Object_list::size_type v_end2 = O2.size();
-  bool splitted = classify_objects( v_end, O.end(), partition_plane, sop,
-                                    std::back_inserter(O1),
-                                    std::back_inserter(O2), depth);
-
-  bool non_efective_split = false;
-  if( !splitted) {
+
+  Vertex_list V1,V2;
+  classify_objects(V, sop, V1, V2);
+
+  Halfedge_list E1,E2;
+  bool splitted = classify_objects(E, sop, E1, E2);
+
+  Halffacet_list F1,F2;
+  splitted = splitted && classify_objects(F, sop, F1, F2);
+
+  if(!splitted) {
     CGAL_NEF_TRACEN("build_kdtree: splitting plane not found");
-    //    if(depth > max_depth)
-    nodes.push_back(Node( parent, nullptr, nullptr, Plane_3(), O));
+    nodes.push_back(Node(V, E, F));
     return &(nodes.back());
-  } else {
-    CGAL_NEF_TRACEN("Sizes " << O1.size() << ", " << O2.size() << ", " << O.size());
-    CGAL_assertion( O1.size() <= O.size() && O2.size() <= O.size());
-    CGAL_assertion( O1.size() + O2.size() >= O.size());
-    non_efective_split = ((O1.size() == O.size()) || (O2.size() == O.size()));
   }
-  if( non_efective_split)
+  auto O_size = V.size() + E.size() + F.size();
+  auto O1_size = V1.size() + E1.size() + F1.size();
+  auto O2_size = V2.size() + E2.size() + F2.size();
+  CGAL_NEF_TRACEN("Sizes " << O1_size << ", " << O2_size << ", " << O_size);
+  CGAL_assertion( O1_size <= O_size && O2_size <= O_size);
+  CGAL_assertion( O1_size + O2_size >= O_size);
+  if((O1_size == O_size) || (O2_size == O_size))
     non_efective_splits++;
   else
     non_efective_splits = 0;
-  if( non_efective_splits > 2) {
+
+  if(non_efective_splits > 2) {
     CGAL_NEF_TRACEN("build_kdtree: non efective splits reached maximum");
-    nodes.push_back(Node( parent, nullptr, nullptr, Plane_3(), O));
+    nodes.push_back(Node(V, E, F));
     return &(nodes.back());
   }
-  nodes.push_back(Node( parent, nullptr, nullptr, partition_plane, Object_list()));
-  Node_handle node = &(nodes.back());
-  node->left_node = build_kdtree( O1, O1.begin()+v_end1, depth + 1, node, non_efective_splits);
-  node->right_node = build_kdtree( O2, O2.begin()+v_end2, depth + 1, node, non_efective_splits);
-  return node;
+
+  Node_handle left_node = build_kdtree(V1, E1, F1, depth + 1);
+  Node_handle right_node = build_kdtree(V2, E2, F2, depth + 1);
+  nodes.push_back(Node(left_node, right_node, construct_splitting_plane(point_on_plane, coord, typename Traits::Kernel::Kernel_tag())));
+  return &(nodes.back());
 }
 
-template <typename Depth>
-bool can_set_be_divided(Object_iterator start, Object_iterator end, Depth depth) {
+bool can_set_be_divided(int depth, typename Vertex_list::size_type size) {
   if(depth >= max_depth)
     return false;
-  if(std::distance(start,end)<2)
+  if(size <= 2)
```

Comment:
I changed the comparison here from `<2` to `<=2` based on the assumption that 1 vertex per tree cell is overkill. However maybe the code elsewhere assumes there can only be one candidate per cell when for example doing the point location?

Reply:
Is this change the culprit for the runtime errors signaled by @sloriot?    I had approved the PR, and it would make sense to not continue working on a PR  by adding further improvements, but to limit it to fixing bugs in the testsuite.

Reply:
@afabri  The change was always in the PR that you approved, I just added the comment about it later in hindsight. That does raise the question of why this is showing up in the testsuite now where it previously passed. Anyway I am not intending to make any further improvements, like you say, limit this to fixing bugs, and make a new PR for anything else. 

When I have some time this weekend I'll investigate, I think I just need to run the testsuite locally, again, but like I say this didn't show up when I ran the testsuite before.